system/jlib/jstream.cpp

```diff
@@ -1383,6 +1444,69 @@ IBufferedSerialInputStream * createBufferedSerialInputStream(MemoryBuffer & sour
     return new CBufferSerialInputStream(source);
 }
 
+ISerialInputStream *createProgressStream(ISerialInputStream *stream, offset_t offset, offset_t len, const char *msg, unsigned periodSecs)
+{
+    class CProgressStream : public CSimpleInterfaceOf<ISerialInputStream>
+    {
+        Linked<ISerialInputStream> stream;
+        offset_t offset = 0;
+        offset_t totalSize = 0;
+        StringAttr msg;
+        void log(double pct)
+        {
+            PROGLOG("%s - %.2f%% complete", msg.get(), pct);
+        }
+        PeriodicTimer periodTimer;
+    public:
+        CProgressStream(ISerialInputStream *_stream, offset_t _offset, offset_t len, const char *_msg, unsigned periodSecs)
+            : stream(_stream), offset(_offset), msg(_msg), periodTimer(periodSecs*1000, true)
+        {
+            totalSize = offset+len;
+        }
+        ~CProgressStream()
+        {
+            offset_t pos = stream->tell();
+            if (pos == totalSize)
+                log(100.0);
+        }
+        // implements ISimpleReadStream
+        virtual size32_t read(size32_t len, void * data) override
+        {
+            size32_t rd = stream->read(len, data);
+            if (periodTimer.hasElapsed())
+            {
+                offset_t pos = stream->tell();
+                if (unlikely(0 == totalSize))
+                    log(100.0);
+                else
+                {
+                    double pct = ((double)pos) / totalSize * 100;
+                    log(pct);
+                }
+            }
+            return rd;
+        }
+        virtual void skip(size32_t sz) override
+        {
+            stream->skip(sz);
+        }
+        virtual void get(size32_t len, void * ptr) override
+        {
+            stream->get(len, ptr);
```

Comment:
this should check..
Given that read and get should cause progress, it's worth commoning up and making skip do too.